crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -0,0 +1,283 @@
+use bevy_utils::HashMap;
+
+use crate::system::{BoxedSystem, Command, IntoSystem};
+use crate::world::{Mut, World};
+// Needed for derive(Component) macro
+use crate::{self as bevy_ecs};
+use bevy_ecs_macros::Resource;
+
+/// Stores systems, so they can be reused and run in an ad-hoc fashion.
+///
+/// Systems are keyed by their [`SystemId`](crate::system::SystemId).
+/// Repeated calls with the same function type will reuse cached state, including for change detection.
+///
+/// Any [`Commands`](crate::system::Commands) generated by these systems (but not other systems), will immediately be applied.
+///
+/// This type is stored as a [`Resource`](crate::system::Resource) on each [`World`], initialized by default.
+/// However, it will likely be easier to use the corresponding methods on [`World`],
+/// to avoid having to worry about split mutable borrows yourself.
+///
+/// # Limitations
+///
+///  - Stored systems cannot be chained: they can neither have an [`In`](crate::system::In) nor return any values.
+///  - Stored systems cannot recurse or nest: they cannot run other systems via
+///    the [`SystemRegistry`](crate::system::SystemRegistry) methods on [`World`] or [`Commands`](crate::system::Commands).
+///  - Exclusive systems cannot be used.
+/// # Examples
+///
+/// You can run a single system directly on the [`World`], while maintaining state.
+/// It's also possible to register two systems at once.
+///
+/// ```rust
+/// # use bevy_ecs::prelude::*;
+/// #[derive(Resource, Default)]
+/// struct Counter(u8);
+///
+/// fn increment(mut counter: Local<Counter>) {
+///    counter.0 += 1;
+///    println!("{}", counter.0);
+/// }
+///
+/// let mut world = World::default();
+/// let counter_one = world.register_system(increment);
+/// let counter_two = world.register_system(increment);
+/// world.run_system_by_id(counter_one); // -> 1
+/// world.run_system_by_id(counter_one); // -> 2
+/// world.run_system_by_id(counter_two); // -> 1
+/// ```
+///
+/// Change detection:
+///
+/// ```rust
+/// # use bevy_ecs::prelude::*;
+/// #[derive(Resource, Default)]
+/// struct ChangeDetector;
+///
+/// let mut world = World::default();
+/// world.init_resource::<ChangeDetector>();
+/// let detector = world.register_system(|change_detector: ResMut<ChangeDetector>| {
+///     if change_detector.is_changed() {
+///         println!("Something happened!");
+///     } else {
+///         println!("Nothing happaned.");
+///     }
+/// });
+///
+/// // Resources are changed when they are first added
+/// let _ = world.run_system_by_id(detector); // -> Something happened!
+/// let _ = world.run_system_by_id(detector); // -> Nothing happened.
+/// world.resource_mut::<ChangeDetector>().set_changed();
+/// let _ = world.run_system_by_id(detector); // -> Something happened!
+/// ```
+
+#[derive(Resource, Default)]
+pub struct SystemRegistry {
+    last_id: u32,
+    systems: HashMap<u32, (bool, BoxedSystem)>,
```

Comment:
Using a `SystemState` enum or struct with a field `initialized` for `(bool, BoxedSystem)` to convey that the `bool` here represents whether or not the system has been initialized would be more clear.

Reply:
`SystemState` is already used, but I like this idea :)

Reply:
I've picked out `RegisteredSystem` as a struct name. I think `CachedSystem` would be equally good, but I have a sneaking suspicion that `CachedSystem` will be used for something else later.